File: 백준/Platinum/13976. 타일 채우기 2/타일 채우기 2.py
# 타일 채우기2
import sys
input = sys.stdin.readline

MOD = 1_000_000_007

# 분할정복 함수
def divide(n):
    if n == 1:
        return [[4, -1],
                [1, 0]]

    sqrt = divide(n//2)

    if n % 2 == 1:
        return multiplyMatrix(multiplyMatrix(sqrt, sqrt), [[4, -1], [1, 0]])
    else:
        return multiplyMatrix(sqrt, sqrt)
        

# 행렬곱 함수
def multiplyMatrix(m1, m2):
    result = [[0,0],
              [0,0]]

    for i in range(2):
        for j in range(2):
            for k in range(2):
                result[i][j] += m1[i][k] * m2[k][j]
            result[i][j] %= MOD

    return result


n = int(input())

if n % 2:
    print(0)
else:
    base = [[3, 0],
            [1, 0]]
    
    divideCnt = n // 2
    result = divide(divideCnt)
    print(multiplyMatrix(result, base)[1][0])
    

File: 백준/Platinum/13976. 타일 채우기 2/타일 채우기 2.java
import java.io.*;

public class Main {
    static int MOD = 1_000_000_007;

    static long[][] divide(long n) {
        if (n == 1) {
            return new long[][] {
                    {4, -1},
                    {1, 0}
            };
        }

        long[][] half = divide(n/2);

        if (n % 2 == 1) {
            return mulMatrix(mulMatrix(half, half), new long[][] {{4, -1}, {1, 0}});
        } else{
            return mulMatrix(half, half);
        }
    }

    static long[][] mulMatrix(long[][] m1, long[][] m2) {
        long[][] result = new long[][]{{0, 0}, {0, 0}};

        for (int i = 0; i < 2; i++) {
            for (int j = 0; j < 2; j++) {
                for (int k = 0; k < 2; k++) {
                    result[i][j] += (m1[i][k] * m2[k][j] + MOD);
                }
                result[i][j] %= MOD;
            }
        }
        return result;
    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        long[][] base = new long[][]{{3, 0}, {1, 0}};
        long n = Long.parseLong(br.readLine());

        if (n % 2 == 1) {
            System.out.println(0);
        } else {
            long powerCnt = n / 2;

            long[][] result = divide(powerCnt);
            System.out.println(mulMatrix(result, base)[1][0]);
        }
    }
}
